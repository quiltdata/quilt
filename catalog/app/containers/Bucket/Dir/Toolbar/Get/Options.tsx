import { basename } from 'path'

import * as React from 'react'
import * as M from '@material-ui/core'

import * as CodeSamples from 'containers/Bucket/CodeSamples'
import * as Buttons from 'containers/Bucket/Download/Buttons'
import GetOptions from 'containers/Bucket/Toolbar/GetOptions'
import type * as Toolbar from 'containers/Bucket/Toolbar'

const useCodeSamplesStyles = M.makeStyles((t) => ({
  code: {
    marginBottom: t.spacing(2),
  },
}))

interface DirCodeSamplesProps {
  className?: string
  bucket: string
  path: string
}

function DirCodeSamples({ className, bucket, path }: DirCodeSamplesProps) {
  const classes = useCodeSamplesStyles()
  const dest = path ? basename(path) : bucket
  const props = { className: classes.code, bucket, path }
  return (
    <div className={className}>
      <CodeSamples.Quilt3List {...props} />
      <CodeSamples.Quilt3Fetch {...props} dest={dest} />
      <CodeSamples.CliList {...props} />
      <CodeSamples.CliFetch {...props} dest={dest} />
    </div>
  )
}

interface DownloadDirProps {
  dirHandle: Toolbar.DirHandle
}

function DownloadDir({ dirHandle }: DownloadDirProps) {
  // TODO: pass selection to Buttons.DownloadDir
  const [downloading, setDownloading] = React.useState(false)
  React.useEffect(() => {
    if (!downloading) return
    setTimeout(() => setDownloading(false), 1000)
  }, [downloading])
  return (
    <Buttons.DownloadDir
      suffix={`dir/${dirHandle.bucket}/${dirHandle.path}`}
      onClick={(event) => {
        event.stopPropagation()
        setDownloading(true)
      }}
      startIcon={downloading ? <M.CircularProgress size={20} /> : undefined}
    >
      Download ZIP (directory)
    </Buttons.DownloadDir>
  )
}

interface OptionsProps {
  handle: Toolbar.DirHandle
  hideCode?: boolean
}

export default function Options({ handle, hideCode }: OptionsProps) {
  const download = <DownloadDir dirHandle={handle} />
  const code = hideCode ? undefined : (
    <DirCodeSamples bucket={handle.bucket} path={handle.path} />
  )

  return <GetOptions download={download} code={code} />
}
