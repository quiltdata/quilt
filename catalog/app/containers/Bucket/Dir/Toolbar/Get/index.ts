export { default as Options } from './Options'
