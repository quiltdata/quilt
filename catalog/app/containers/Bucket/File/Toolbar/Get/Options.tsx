import { basename } from 'path'

import * as React from 'react'
import * as M from '@material-ui/core'

import * as CodeSamples from 'containers/Bucket/CodeSamples'
import * as Buttons from 'containers/Bucket/Download/Buttons'
import GetOptions from 'containers/Bucket/Toolbar/GetOptions'

import type * as Toolbar from 'containers/Bucket/Toolbar'

const useFileCodeSamplesStyles = M.makeStyles((t) => ({
  code: {
    marginBottom: t.spacing(2),
  },
}))

interface FileCodeSamplesProps {
  className?: string
  bucket: string
  path: string
}

export function FileCodeSamples({ className, bucket, path }: FileCodeSamplesProps) {
  const classes = useFileCodeSamplesStyles()
  const apiDest = basename(path)
  const cliDest = ''
  const props = { className: classes.code, bucket, path }
  return (
    <div className={className}>
      <CodeSamples.Quilt3Fetch {...props} dest={apiDest} />
      <CodeSamples.CliFetch {...props} dest={cliDest} />
    </div>
  )
}

interface OptionsProps {
  handle: Toolbar.FileHandle
  hideCode?: boolean
}

export default function Options({ handle, hideCode }: OptionsProps) {
  const download = <Buttons.DownloadFile fileHandle={handle} />
  const code = hideCode ? undefined : (
    <FileCodeSamples bucket={handle.bucket} path={handle.key} />
  )

  return <GetOptions download={download} code={code} />
}
