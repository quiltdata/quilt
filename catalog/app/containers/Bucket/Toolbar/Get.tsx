import * as React from 'react'

import { GetAppOutlined as IconGetAppOutlined } from '@material-ui/icons'

import * as Buttons from 'components/Buttons'

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  label?: string
}

export default function Button({ label = 'Get files', ...props }: ButtonProps) {
  return <Buttons.WithPopover icon={IconGetAppOutlined} label={label} {...props} />
}
