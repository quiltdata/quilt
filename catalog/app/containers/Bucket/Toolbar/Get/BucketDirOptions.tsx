import * as React from 'react'
import * as M from '@material-ui/core'

import { Tabs } from 'components/Dialog'
import * as Buttons from 'containers/Bucket/Download/Buttons' // TODO: move in this dir

import type { DirHandle } from '../types'

import { DirCodeSamples } from './BucketCodeSamples'

interface DownloadDirProps {
  dirHandle: DirHandle
}

function DownloadDir({ dirHandle }: DownloadDirProps) {
  return (
    <Buttons.DownloadDir suffix={`dir/${dirHandle.bucket}/${dirHandle.path}`}>
      Download ZIP (directory)
    </Buttons.DownloadDir>
  )
}

const useStyles = M.makeStyles((t) => ({
  download: {
    width: t.spacing(40),
  },
  code: {
    width: t.spacing(80),
  },
}))

interface BucketOptionsProps {
  handle: DirHandle
  hideCode?: boolean
}

export default function BucketDirOptions({ handle, hideCode }: BucketOptionsProps) {
  const classes = useStyles()
  const download = React.useCallback(
    () => ({
      className: classes.download,
      label: 'Download',
      panel: <DownloadDir dirHandle={handle} />,
    }),
    [classes.download, handle],
  )
  const code = React.useCallback(
    () => ({
      className: classes.code,
      label: 'Code',
      panel: <DirCodeSamples bucket={handle.bucket} path={handle.path} />,
    }),
    [classes.code, handle],
  )
  const tabs = React.useMemo(
    () => (hideCode ? [download()] : [download(), code()]),
    [code, download, hideCode],
  )
  return <Tabs>{tabs}</Tabs>
}
