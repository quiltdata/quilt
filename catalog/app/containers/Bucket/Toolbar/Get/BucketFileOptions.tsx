import * as React from 'react'

import BucketOptions from '../../Download/BucketOptions'

import type { FileHandle } from '../types'

interface BucketFileOptionsProps {
  handle: FileHandle
  hideCode?: boolean
}

export default function BucketFileOptions({ handle, hideCode }: BucketFileOptionsProps) {
  return <BucketOptions handle={handle} hideCode={hideCode} />
}
