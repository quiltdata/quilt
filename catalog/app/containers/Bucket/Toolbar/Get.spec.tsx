import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Get from './Get'

jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  IconButton: ({ className, children, onClick }: any) => (
    <button role="button" {...{ className, children, onClick }} />
  ),
  Button: ({ className, children, onClick }: any) => (
    <button role="button" {...{ className, children, onClick }} />
  ),
}))

describe('containers/Bucket/Toolbar/Get', () => {
  it('should render with default label', () => {
    const { container } = render(<Get className="custom-class">Hello, Popover!</Get>)
    expect(container).toMatchSnapshot()
  })

  it('should render with custom label', () => {
    const { container } = render(<Get label="Custom Get Label">Hello, Popover!</Get>)
    expect(container).toMatchSnapshot()
  })

  it('should render children when popover is opened', () => {
    const { container, getByRole } = render(<Get>Hello, Popover!</Get>)
    const button = getByRole('button')
    act(() => {
      fireEvent.click(button)
    })
    expect(container).toMatchSnapshot()
  })
})
