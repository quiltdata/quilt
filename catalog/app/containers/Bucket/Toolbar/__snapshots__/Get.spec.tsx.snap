// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/Toolbar/Get should render children when popover is opened 1`] = `
<div>
  <div
    class="makeStyles-root-9"
  >
    <button
      role="button"
    >
      Get files
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-11"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      class="MuiPaper-root makeStyles-popup-10 MuiPaper-elevation4 MuiPaper-rounded"
    >
      Hello, Popover!
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar/Get should render with custom label 1`] = `
<div>
  <div
    class="makeStyles-root-5"
  >
    <button
      role="button"
    >
      Custom Get Label
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-7"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar/Get should render with default label 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <button
      class="custom-class"
      role="button"
    >
      Get files
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-3"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;
